overlord/snapstate/snapmgr.go

```diff
@@ -434,11 +441,26 @@ func (m *SnapManager) ensureRefreshes() error {
 	}
 
 	// store attempts in memory so that we can backoff a
-	m.lastRefreshAttempt = time.Now()
-	updated, tasksets, err := AutoRefresh(m.state)
-	if err != nil {
-		return err
-	}
+	delta := timeutil.Next(refreshSchedule, lastRefresh)
+	m.nextRefresh = time.AfterFunc(delta, func() {
+		m.lastRefreshAttempt = time.Now()
+		m.state.Lock()
+		updated, tasksets, err := AutoRefresh(m.state)
+		m.state.Unlock()
```

Comment:
the unlocking looks wrong here, we need to make the Change based on tasksets while we still old the same lock